From c5eb7faeb53a4d4a3ee3b7d3b5694c16dc23507c Mon Sep 17 00:00:00 2001
From: Ola Jeppsson <ola@adapteva.com>
Date: Tue, 8 Nov 2016 18:16:30 +0100
Subject: [PATCH 32/33] libgcc:epiphany: Use registers marked in epiphany.md
 for builtins

epiphany.md does not mark r15 as clobbered for __modsi3, nor __umodsi3.
Replace those uses with a unused register that are marked as clobbered.

libgcc/
	* config/epiphany/modsi3.S (__modsi3): Use r18 instead of r15.
	* config/epiphany/umodsi3.S (__umodsi3): Use r17 instead of r15.

Signed-off-by: Ola Jeppsson <ola@adapteva.com>
---
 libgcc/ChangeLog.epiphany        |  5 +++++
 libgcc/config/epiphany/modsi3.S  | 10 +++++-----
 libgcc/config/epiphany/umodsi3.S | 10 +++++-----
 3 files changed, 15 insertions(+), 10 deletions(-)

diff --git a/libgcc/ChangeLog.epiphany b/libgcc/ChangeLog.epiphany
index adea2d3b459..18b0aaf70a1 100644
--- a/libgcc/ChangeLog.epiphany
+++ b/libgcc/ChangeLog.epiphany
@@ -1,3 +1,8 @@
+2016-11-08  Ola Jeppsson  <ola@adapteva.com>
+
+	* config/epiphany/modsi3.S (__modsi3): Use r18 instead of r15.
+	* config/epiphany/umodsi3.S (__umodsi3): Use r17 instead of r15.
+
 2015-09-03  Ola Jeppsson  <ola@adapteva.com>
 
 	* config/epiphany/crti.S (_init) (_fini): Respect
diff --git a/libgcc/config/epiphany/modsi3.S b/libgcc/config/epiphany/modsi3.S
index 6d63721b619..619960466ac 100644
--- a/libgcc/config/epiphany/modsi3.S
+++ b/libgcc/config/epiphany/modsi3.S
@@ -38,12 +38,12 @@ SYM(__modsi3):
 	movgt r1,r3
 	movt r2,0xa000 ; 0xa0000000
 	orr r3,r2,r0
-	lsr r15,r0,16
-	movt r15,0xa800
-	movne r3,r15
+	lsr r18,r0,16
+	movt r18,0xa800
+	movne r3,r18
 	lsr r16,r2,2 ; 0x28000000
-	and r15,r3,r16
-	fadd r12,r3,r15
+	and r18,r3,r16
+	fadd r12,r3,r18
 	 orr r3,r2,r1
 	lsr r2,r1,16
 	movt r2,0xa800
diff --git a/libgcc/config/epiphany/umodsi3.S b/libgcc/config/epiphany/umodsi3.S
index abc3c66a65e..16355d01332 100644
--- a/libgcc/config/epiphany/umodsi3.S
+++ b/libgcc/config/epiphany/umodsi3.S
@@ -33,12 +33,12 @@ SYM(__umodsi3):
 	mov r2,5
 	lsl r2,r2,29 ; 0xa0000000
 	orr r3,r2,r0
-	lsr r15,r0,16
-	movt r15,0xa800
-	movne r3,r15
+	lsr r17,r0,16
+	movt r17,0xa800
+	movne r3,r17
 	lsr r16,r2,2 ; 0x28000000
-	and r15,r3,r16
-	fadd r12,r3,r15
+	and r17,r3,r16
+	fadd r12,r3,r17
 	 orr r3,r2,r1
 	lsr r2,r1,16
 	movt r2,0xa800
-- 
2.31.1

